### ISSUE TITLE
Cart item ids might be causing more problems than expected?

### ISSUE THREAD

**Issue by rkeeves** [saucelabs/sample-app-web#118]
## TLDR

 PR #115 introduced a new private route `**/inventory-long.html`. Now 'bug looking' behaviors can be triggered more easily.

_The PR didn't introduce these bugs. It just made them more 'visible'._

I know there are other PRs in plans, which are going to do refactors. I just wanted to bring some symptoms up.

I don't know the specification of this software, and given the software's purpose (browser-based testing demo), maybe some or all of these behaviors are  actual features?! _(I mean this app's primary purpose is to be tested to evaluate some testing tool/service etc.)_

## Symptom A - Counter goes to 69

1. Login as `standard_user`
1. Goto `https://www.saucedemo.com/inventory-item.html?id=foo`
1. Click Add, click remove 69 times
1. See counter at 69
1. Goto Cart
1. See error

## Symptom B - InventoryItem page allows adding/removing unknown item

1. Login as `standard_user`
1. Goto `https://www.saucedemo.com/inventory-item.html?id=69`
1. Click Add
1. Click Remove
1. Click Add
1. Goto Cart
1. See error

## Symptom C - InventoryItem does not display known item

1. Login as `standard_user`
1. Navigate to `https://www.saucedemo.com/inventory-long.html`
1. Visit last item via link
1. See InventoryItem page not displaying the item details and instead show 'Item Not Found'

## Symptom D - Cart throws error

1. Login as `standard_user`
1. Navigate to `https://www.saucedemo.com/inventory-long.html`
1. Add last item
1. Goto Cart
1. See error

## Symptom E - CheckOutStepTwo throws error

1. Login as `standard_user`
1. Navigate to `https://www.saucedemo.com/inventory-long.html`
1. Add last item
1. Navigate to `https://www.saucedemo.com/checkout-step-two.html`
1. See error

## Symptom F - React components' and localstorage's state diverges (missing updates?!)

1. Login as `standard_user`
1. Navigate to `https://www.saucedemo.com/inventory.html`
1. Add last item
1. Reset app state
1. Still see 'Remove' buttons
1. See counter 0

## Background

There's no single root cause.

Different things together cause these problems:
- nulls and invalid ids can be written into local storage
- `shopping-cart.js` uses observer pattern which causes components to not be in sync
- some React components and pages index into a specific array (constant) unsafely
- queryParam and parseInt can result in weird values (which will reach `shopping-cart.js` and then localstorage)
- all Pages except for `InventoryItem` have no way of graceful error handling (related to cart item id array)
- there are two different sets of items (yep, I know the refactoring is currently under way, but now it is easy to trigger strange behavior)
- ui and state is not decoupled, so it is a bit hard to judge what is a real bug and what is a 'built in' bug

**Comment by diemol** [comment]
Thank you for taking the time to raise this issue. 

This application is meant to be a sample and display happy paths and bugs to allow users to catch bugs through their testing. There are several places where bugs exist; this is intended, as this is simply a demo application.

**Comment by rkeeves** [comment]
Yep, I'll go away:

Here's [the spec](https://github.com/saucelabs/sample-app-web/commit/90ecaa2b59d07da4f7c4bad83cc99b6b99ed4ef9).

Here's [where NaN is assigned](https://github.com/saucelabs/sample-app-web/blob/2bf78294911ddf05070b5e38e79061eda5092bfe/src/pages/InventoryItem.jsx#L44).

```
parseInt('') is NaN

NaN >= 0 && 6 > NaN is false

JSON.parse(JSON.stringify([NaN])) is [null]

[null].indexOf(NaN) is -1
```
